lib/fastlane/plugin/wpmreleasetoolkit/helper/android/android_localize_helper.rb

```diff
@@ -299,6 +299,29 @@ def self.merge_xml_documents(all_xmls)
         #
         def self.apply_substitutions(tag)
           tag.content = tag.content.gsub('...', '…')
+
+          # Typography en-dash
+          if tag.content.include?('-')
+            matcher = tag.content.match(/.*(\d+\s*)-(\s*\d+).*/)
+            index = 1
+            while index < matcher.length do
+              # Make sure that if there is no space before digit there isn't
+              # one on the left either -- since we don't want to consider
+              # "1 2 -3" as a range from 2 to 3
+              isNegativeNumber = (matcher[index + 1][0] != ' ') && (matcher[index][matcher[index].length - 1] == ' ')
+
+              # Do the substitution -- match the text block instead of the single dash to make sure that
+              # if there are more dashes in the tag, only the right ones get substituted.
+              puts "Original: #{tag.content} - #{isNegativeNumber}"
+              tag.content = tag.content.gsub(
+                "#{matcher[index]}-#{matcher[index + 1]}",
+                "#{matcher[index]}\u2013#{matcher[index + 1]}"
+                ) unless isNegativeNumber
+              puts "Updated: #{tag.content}"
+
+              index = index + 2
+            end unless matcher.nil?
+          end
```

Comment:
Here `matcher.length` does not do what you think it does.
Instead `matcher` is a `MatchData` (the **first and only** match that `.match` returned in previous line) and `matcher.length` contains the number of capture groups. So on the string `'abc 1-2 def'`, matcher[0] is the `1` and `matcher[1]` is the `2` because those are the substrings captured by the parentheses in your regex.

I think all this could be replaced by either using `scan` (to return all occurrences found using a regex instead of the first one like with `.match`), or looking at ruby doc to see if there is not a variant of `.match` that would return all the matches not the first one.

Another, maybe better, way shorter, and more elegant solution would be to use the block variant of `gsub`. Not tested but I think a code like this one could be enough to replace you whole `while` loop:

```ruby
tag.content.gsub(/(\d+\s*)-(\s*\d+)/) do |str|
  match = Regexp.last_match # of type `MatchData`. match[0] == str == whole match, match[1] = 1st capture group (left part of the range), match[2] = second capture group (right part of the range)
  isNegativeNumber = match[2][0] != ' ' && match[1][-1] == ' ' # if right part of range is not a space but left part of range ends with space, it's not a range after all
  isNegativeNumber ? str : "#{match[1]}\u2013#{match[2]}"
end
```

Reply:
> Instead matcher is a MatchData (the first and only match that .match returned in previous line) and matcher.length contains the number of capture groups. So on the string 'abc 1-2 def', matcher[0] is always the whole matched string, and matcher[1] is the "1" while matcher[2] is the "2" – because those are the substrings captured by the parentheses in your regex.

Yep, this is exactly what I expected 🤔 👍

I love your better, way shorter and more elegant solution, but unluckily it doesn't work. I think that it doesn't catch the last dash in the string, so it basically always fails when there's only one dash (probably the most frequent case).

Reply:
Sorry, I was wrong. The problem was with the assignment. I thought `gsub!` would do the substitutions in the original `tag.content`, though it doesn't seem to be the case.

Reply:
Yep, see my previous comment :-)

Reply:
Ah, right! 👌 
So `tag.content = tag.content.gsub(…)` should work then :)

Reply:
Yes, it works. But isn't `gsub!` supposed to substitute in the original variable? 🤔

Reply:
Yep, it should indeed 🧐  not sure why it doesn't 🤔